### ISSUE TITLE
Unexpected / unintuitive behaviour of tuple(DeviceArray)

### ISSUE THREAD

**Issue by szepi1991** [google/jax#3330]
`tuple(DeviceArray)` evaluates the DeviceArray (at least when using xla) and returns them as np arrays. 

Example:
```python
ar = jnp.ones(shape=3, dtype=jnp.float32)    # type: jax.interpreters.xla.DeviceArray,
tuple(ar)
```
produces a tuple of `numpy.float32: (1.0, 1.0, 1.0)`, while
```python
tuple(ar[i] for i in range(ar.shape[0]))
```
produces a tuple of DeviceArrays.

It feels like `tuple(ar)` returning a tuple of DeviceArrays would make sense. Is there some drawback to this?

Furthermore, when jitting vs not-jitting the following:
```python
def tuple_split(ar):
  return tuple(ar)
```
I do get different return types like so:
```python
print("orig:", tuple_split(jnp.ones(shape=3, dtype=jnp.float32)))
> orig: (1.0, 1.0, 1.0)
print("jit:", jit_tuple_split(jnp.ones(shape=3, dtype=jnp.float32)))
> jit: (DeviceArray(1., dtype=float32), DeviceArray(1., dtype=float32), DeviceArray(1., dtype=float32))
```
which I also didn't expect.

xla tuple implementation: https://github.com/google/jax/blob/master/jax/interpreters/xla.py#L1044

**Comment by mattjj** [comment]
Great question! Take a look at #965 for an explanation of why things are currently implemented this way.

I suppose we could make `tuple(device_array)` return a tuple of DeviceArrays _even if_ we kept `DeviceArray.__iter__` implemented via bouncing to the host, i.e. by writing it like `map(device_put, self._value.__iter__())`.

@hawkinsp any thoughts to add?

**Comment by szepi1991** [comment]
I see, thank you for sharing that, that makes sense.

My understanding is that your suggestion would copy contents to the host, iterate there, then copy back onto the device. Is my intuition correct that copying back and forth may be slow, but the idea is that it's still faster than taking slices on the device, at least when there are many slices?

I'm now somewhat torn since I could also reasonably expect that __iter__ and tuple() do the same thing.

Is it fairly usual that the original and the jitted function have different return types like in my above example? Can you simply construct another example (other than using __iter__ or tuple)?

**Comment by shoyer** [comment]
From an API consistency perspective, I think it would be best to always return DeviceArray objects.

Iterating via repeated indexing vs copying to the host strikes me as an optimization detail that shouldn’t need to impact the public API. I would actually guess that there could be some cases where indexing via repeated indexing _is_ faster, namely for arrays where each element along the first axis contains a very large number of values.

**Comment by szepi1991** [comment]
The following is not entirely relevant for the original issue, though it's related to the optimization detail bit so I'll share.

We did encounter a situation where the copying behaviour hurt us, though arguable we were not using jax right. We were doing:
```
# BAD
new_rng_key, subkey = jrandom.split(rng_key)
output = jitted_fn(subkey, input)
```
whereas we should have done:
```
new_rng_key, output = jitted_fn(rng_key, input)   # split happens inside
```
In the first case, on GPU/TPU the tuple expansion results in a copy to host; then the next line copies the data back to the device. This makes it a LOT slower than the second case where the computation simply continues on the device.

**Comment by mattjj** [comment]
I think this is a duplicate of #1583. Let's close this issue to consolidate the discussion there. (Sorry for the fragmentation that's already happened, but at least this issue helped us restart the discussion!)

@szepi1991 that's a really good example! Actually I like your "BAD" version because it doesn't require threading the PRNG key back out, which is a goal of the JAX PRNG API. See the [PRNG design doc](https://github.com/google/jax/blob/master/design_notes/prng.md).

But indeed due to the problem we're discussing in this thread it means bouncing the result to the host, which is problematic. In other words, the `tuple(DeviceArray)` problem is preventing the use of our preferred PRNG pattern!
